# The beginning of the journey

## Day 1

### Five steps in omics data analysis

Generally speaking, I consider omics data analysis consisting of five steps:

(1) Data generation and collection. Why should we do an omics data analysis? Usually it's not for fun, though it indeed has a lot of fun.

Scientific studies usually aim to test a hypothesis, or generate new hypotheses. Usually it's a combination of both. That is to test a hypothesis, meanwhile to generate new hypotheses.

Depending on the hypothesis to test or generate, we need different data to answer.

(2) Data quality check and cleaning

(3) Data processing.

Sometimes, you need Supercomputerome, such as ([Computerome](https://www.computerome.dk/))

(4) Exploratory data analysis and modelling

(5) Visualization and reporting.

**Exercise:**

1.  Find out what is [Computerome](www.computerome.dk/)
2.  Which step(s) of omics data analysis is less familiar to you, and which may be a challenge for you?

### Getting started with R

#### R and RStudio

#### CRAN and Bioconductor

-   R package repository. For each distribution and dependency (compatibility) management

-   CRAN vs. Bioconductor

-   CRAN: R core team; general purpose statistics

-   Bioconductor: reviewed; for biology

### Statistics

#### Why do we need statistics?

#### Classification and regression

## Day 2

### Recap

-   What is **reproducibility** in omics data analysis? Which factors could affect reproducibility?

-   R and RStudio

-   CRAN and Bioconductor

### R: loading a package, and get help

library and package.

```{r day 2 load a library, eval=FALSE}
library(gprofiler2)
library(ggplot2)
library(dplyr)
```

### R: the basics

#### Arithmatic operation

```{r eval=FALSE}
2 + 3 * 5       # Note the order of operations.
(2 + 3) * 5 
log(10)         # Natural logarithm with base e
5^2             # 5 raised to the second power
3/2             # Division
sqrt(16)        # Square root
abs(3-7)        # Absolute value of 3-7
pi              # The number
exp(2)          # exponential function
28%%6           # Modulo
```

**Exercise:**

-   Calculate

    -   `r round(runif(1, min = -0, max = 200))` plus `r round(runif(1, min = -0, max = 200))`

    -   `r round(runif(1, min = -0, max = 200))` minus `r round(runif(1, min = -0, max = 200))`

    -   `r round(runif(1, min = -0, max = 200))` times `r round(runif(1, min = -0, max = 200))`

    -   the average of `r round(runif(1, min = -0, max = 200))` and `r round(runif(1, min = -0, max = 200))`

    -   the sum of `r round(runif(1, min = -0, max = 200))` and `r round(runif(1, min = -0, max = 200))`

    -   the difference between `r round(runif(1, min = -0, max = 200))` and `r round(runif(1, min = -0, max = 100))`

    -   the ratio between `r round(runif(1, min = -0, max = 200))` and `r round(runif(1, min = -0, max = 200))`

    -   `r round(runif(1, min = -0, max = 50))` multiply `r round(runif(1, min = -0, max = 200))`

-   Execute `log(45)` and `log2(45)` and compare the difference. Note: you can get help using `help(log)` and `help(log2)`

-   A gene expression: CO (mean: 12) TRET (mean: 50) upregulated; Log2 fold change: `r log2(50/12)`

-   B gene expression: CO (mean: 50) TEAT (mean: 12) downregulated; Log2 fold change: `r log2(12/50)`

-   log computes logarithms, by default natural logarithms, log10 computes common (i.e., base 10) logarithms, and log2 computes binary (i.e., base 2) logarithms. The general form log(x, base) computes logarithms with base base.

```{r Answer, include=FALSE}
141 + 46
117 - 19
100 * 105
(193 + 44) / 2

```

#### Variable assignment

```{r, eval=FALSE}
# Assign the value 42 to x
x <- 68
# Print out the value of the variable x
x
```

```{r, eval=FALSE}
# Assign the value 5 to the variable my_apples
my_apples <- 5
# Print out the value of the variable my_apples
my_apples
```

```{r, eval=FALSE}
# Assign a value to the variables my_apples and my_oranges
my_apples <- 5
my_oranges <- 6
# Add these two variables together
my_apples + my_oranges
# Create the variable my_fruit
my_fruit <- my_apples + my_oranges
```

#### Variable type

```{r, eval=FALSE}
# Change my_numeric to be 42
my_numeric <- 42
# Change my_character to be "universe"
my_character <- "universe"
# Change my_logical to be FALSE
my_logical <- FALSE
```

```{r, eval=FALSE}
# Check class of my_numeric
class(my_numeric)
# Check class of my_character
class(my_character)
# Check class of my_logical
class(my_logical)
```

<!-- exercise -->

**Exercise:**

Test out the following code, and figure out what do they mean?

```{r eval=FALSE}
logic_1 <- TRUE
logic_2 <- "TRUE"
logic_3 <- "True"
```

```{r eval=FALSE}
2 == 3
logic_1 == logic_2
logic_2 == logic_3
logic_1 == logic_3
logic_1 == 1
```

```{r eval=FALSE}
1 < 2
1 > 2
1 == 2
1 != 2 # (not equal to)
```

#### Vector

```{r, eval=FALSE}
numeric_vector <- c(1, 10, 49)
character_vector <- c("a", "b", "c")

# Complete the code for boolean_vector
boolean_vector <-c(TRUE,FALSE,TRUE)

# Poker winnings from Monday to Friday
poker_vector <- c(140, -50, 20, -120, 240)

# Roulette winnings from Monday to Friday
roulette_vector <- c(-24, -50, 100, -350, 10)

# Assign days as names of poker_vector
names(poker_vector) <- c("Monday", "Tuesday", "Wednesday", "Thursday", "Friday")

# Assign days as names of roulette_vector
names(roulette_vector)<-c("Monday", "Tuesday", "Wednesday", "Thursday", "Friday")
```

```{r, eval=FALSE}
A_vector <- c(1, 2, 3)
B_vector <- c(4, 5, 6)

# Take the sum of A_vector and B_vector
total_vector <- A_vector + B_vector

C_vector <- c(A_vector, A_vector)
D_vector <- c(A_vector, B_vector)

A_vector * B_vector
A_vector + 1 
```

Recap

-   variable

-   class: numeric, character, logical

## Day 3

### Recap

1.  Read the following paragraph clipped from `help(log)`, and reflect what does it mean?

    log computes logarithms, by default natural logarithms, log10 computes common (i.e., base 10) logarithms, and log2 computes binary (i.e., base 2) logarithms. The general form log(x, base) computes logarithms with base base.

2.  Assuming **FLG** (filaggrin, a gene coding an important skin barrier protein) has an average expression (count) of 500 in healthy control, and 200 in the disease group. How to calculate `log2 fold change (log2 FC)`? Think both mathematically, and in R.

3.  If FLG is downregulated in atopic dermatitis skin disease, what does it mean?

4.  What is vector recycling in R? What could be the consequence?

5.  What do these signs mean in R?

        != 
        ==
        <
        >
        ?
        >=
        <=
        <-

### Vector, named vector, vector comparison

In an experiment, we measured 8 biomarkers and stored theirs values in two variables (CO and TRET). How can I compare these 8 biomarkers in two conditions?

Use this code to generate CO and TRET

```{r vector comparison, result = 'hide'}
n <- 8
CO <- round(runif(n, 0, 100))
TRET <- round(runif(n, 0, 100))
names(CO) <- paste("biomarker", 1: n)
names(TRET) <- paste("biomarker", 1: n)
```

```{r}
CO
TRET
```

```{r include=FALSE}
CO < TRET
TRET - CO
```

### Matrix (excerpted from [@akalin2020a])

```{r eval=FALSE}
?matrix
```

A matrix refers to a numeric array of rows and columns. You can think of it as a stacked version of vectors where each row or column is a vector. One of the easiest ways to create a matrix is to combine vectors of equal length using `cbind()`, meaning 'column bind'.\index{R Programming Language!matrix}

```{r matrices}
x<-c(1,2,3,4)
y<-c(4,5,6,7)
m1<-cbind(x,y);m1
t(m1)                # transpose of m1
dim(m1)              # 2 by 5 matrix
```

You can also directly list the elements and specify the matrix:

```{r matrix2}
m2<-matrix(c(1,3,2,5,-1,2,2,3,9),nrow=3)
m2
```

Matrices and the next data structure, **data frames**, are tabular data structures. You can subset them using `[]` and providing desired rows and columns to subset. Figure \@ref(fig:slicingDataFrames) shows how that works conceptually.

### Data frames (excerpted from [@akalin2020a])

A data frame is more general than a matrix, in that different columns can have different modes (numeric, character, factor, etc.). A data frame can be constructed by the `data.frame()` function. For example, we illustrate how to construct a data frame from genomic intervals or coordinates.\index{R Programming Language!data frame}

```{r dfcreate}
chr <- c("chr1", "chr1", "chr2", "chr2")
strand <- c("-","-","+","+")
start<- c(200,4000,100,400)
end<-c(250,410,200,450)
mydata <- data.frame(chr,start,end,strand)
#change column names
names(mydata) <- c("chr","start","end","strand")
mydata # OR this will work too
mydata <- data.frame(chr=chr,start=start,end=end,strand=strand)
mydata
```

There are a variety of ways to extract the elements of a data frame. You can extract certain columns using column numbers or names, or you can extract certain rows by using row numbers. You can also extract data using logical arguments, such as extracting all rows that have a value in a column larger than your threshold.

```{r dfSlice}
mydata[,2:4] # columns 2,3,4 of data frame
mydata[,c("chr","start")] # columns chr and start from data frame
mydata$start # variable start in the data frame
mydata[c(1,3),] # get 1st and 3rd rows
mydata[mydata$start>400,] # get all rows where start>400
```
